File: frontend/src/services/challenges.ts
import { api } from '@/services/api';
import type { PublicUserResponse } from '@/services/auth';
import type { MemeResponse } from '@/services/memes';
import { appendImageToFormData } from '@/utils/multipartImage';

export type ChallengeType = 'intra_community' | 'community_vs_community' | 'open' | 'duel';
export type ChallengeStatus = 'setup' | 'active' | 'evaluated';

export interface ChallengeSideResponse {
  id: string;
  name: string;
  community_id: string | null;
  member_ids: string[];
  participant_count: number;
  score: number | null;
}

export interface ChallengeResponse {
  id: string;
  // Null for `open` and `duel` challenges — both are platform-level, no community.
  community_id: string | null;
  opponent_community_id: string | null;
  hashtag: string | null;
  creator: PublicUserResponse;
  // `duel` only: the challenged friend, set from proposal even before they've accepted.
  invitee_id: string | null;
  invitee: PublicUserResponse | null;
  title: string;
  challenge_type: ChallengeType;
  status: ChallengeStatus;
  start_time: string;
  end_time: string;
  winning_side_id: string | null;
  sides: ChallengeSideResponse[];
}

export interface ChallengeSubmissionResponse {
  id: string;
  side_id: string;
  submitter: PublicUserResponse;
  meme: MemeResponse;
  created_at: string;
}

export interface ChallengeResultsResponse {
  challenge: ChallengeResponse;
  submissions: ChallengeSubmissionResponse[];
}

export interface ChallengeSideSetup {
  name: string;
  member_ids: string[];
}

export function createChallengeRequest(
  communityId: string,
  payload: { title: string; start_time: string; end_time: string; sides: ChallengeSideSetup[] }
) {
  return api.post<ChallengeResponse>(`/communities/${communityId}/challenges`, payload);
}

export function proposeVsChallengeRequest(
  communityId: string,
  opponentCommunityId: string,
  payload: { title: string; start_time: string; end_time: string }
) {
  return api.post<ChallengeResponse>(
    `/communities/${communityId}/challenges/vs/${opponentCommunityId}`,
    payload
  );
}

export function acceptChallengeRequest(communityId: string, challengeId: string) {
  return api.post<ChallengeResponse>(
    `/communities/${communityId}/challenges/${challengeId}/accept`
  );
}

export function declineChallengeRequest(communityId: string, challengeId: string) {
  return api.delete(`/communities/${communityId}/challenges/${challengeId}/decline`);
}

export function getCommunityChallengesRequest(communityId: string) {
  return api.get<ChallengeResponse[]>(`/communities/${communityId}/challenges`);
}

export function getChallengeRequest(communityId: string, challengeId: string) {
  return api.get<ChallengeResponse>(`/communities/${communityId}/challenges/${challengeId}`);
}

export function submitToChallengeRequest(
  communityId: string,
  challengeId: string,
  memeId: string
) {
  return api.post<ChallengeSubmissionResponse>(
    `/communities/${communityId}/challenges/${challengeId}/submissions`,
    undefined,
    { params: { meme_id: memeId } }
  );
}

export function getChallengeResultsRequest(communityId: string, challengeId: string) {
  return api.get<ChallengeResultsResponse>(
    `/communities/${communityId}/challenges/${challengeId}/results`
  );
}

// --- Flat (community-less) endpoints — used by duels and, incidentally, open challenges,
// neither of which has a communityId to put in the URL. ---

export function getChallengeFlatRequest(challengeId: string) {
  return api.get<ChallengeResponse>(`/challenges/${challengeId}`);
}

export function getChallengeResultsFlatRequest(challengeId: string) {
  return api.get<ChallengeResultsResponse>(`/challenges/${challengeId}/results`);
}

export async function createAndSubmitToChallengeRequest(
  challengeId: string,
  image: { uri: string; name: string; type: string },
  caption?: string
) {
  const form = new FormData();
  await appendImageToFormData(form, 'image', image);
  if (caption) form.append('caption', caption);

  return api.post<ChallengeSubmissionResponse>(
    `/challenges/${challengeId}/submissions`,
    form,
    { headers: { 'Content-Type': undefined } }
  );
}

export function proposeDuelRequest(
  opponentId: string,
  payload: { title: string; start_time: string; end_time: string }
) {
  return api.post<ChallengeResponse>(`/challenges/duels/${opponentId}`, payload);
}

export function acceptDuelRequest(challengeId: string) {
  return api.post<ChallengeResponse>(`/challenges/duels/${challengeId}/accept`);
}

export function declineDuelRequest(challengeId: string) {
  return api.delete(`/challenges/duels/${challengeId}/decline`);
}

// --- Open challenges + cross-community "mine" list (Phase 20/18 frontend). ---

export interface OpenChallengeSideSetup {
  name: string;
}

export function createOpenChallengeRequest(payload: {
  title: string;
  hashtag: string;
  start_time: string;
  end_time: string;
  sides: OpenChallengeSideSetup[];
}) {
  return api.post<ChallengeResponse>('/challenges/open', payload);
}

export function listOpenChallengesRequest() {
  return api.get<ChallengeResponse[]>('/challenges/open');
}

export function joinOpenChallengeRequest(challengeId: string, sideId: string) {
  return api.post<ChallengeResponse>(`/challenges/${challengeId}/join`, { side_id: sideId });
}

export function listMyChallengesRequest() {
  return api.get<ChallengeResponse[]>('/challenges/mine');
}

File: backend/app/schemas/challenges.py
import datetime
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.models.challenge import ChallengeStatus, ChallengeType
from app.schemas.auth import PublicUserOut
from app.schemas.memes import MemeOut


class ChallengeSideSetup(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    member_ids: list[uuid.UUID] = Field(min_length=1)


class ChallengeCreate(BaseModel):
    title: str = Field(min_length=1, max_length=150)
    start_time: datetime.datetime
    end_time: datetime.datetime
    sides: list[ChallengeSideSetup] = Field(min_length=2)


class ChallengeProposalCreate(BaseModel):
    title: str = Field(min_length=1, max_length=150)
    start_time: datetime.datetime
    end_time: datetime.datetime


class OpenChallengeSideSetup(BaseModel):
    """No `member_ids` — an open challenge's sides are filled by self-service join."""

    name: str = Field(min_length=1, max_length=100)


class OpenChallengeCreate(BaseModel):
    title: str = Field(min_length=1, max_length=150)
    # Normalized server-side; a challenge reserves its tag exclusively.
    hashtag: str = Field(min_length=1, max_length=100)
    start_time: datetime.datetime
    end_time: datetime.datetime
    sides: list[OpenChallengeSideSetup] = Field(min_length=2)


class ChallengeJoin(BaseModel):
    side_id: uuid.UUID


class DuelCreate(BaseModel):
    title: str = Field(min_length=1, max_length=150)
    start_time: datetime.datetime
    end_time: datetime.datetime


class ChallengeSideOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    community_id: uuid.UUID | None
    # Always empty for an open challenge — its roster is unbounded, use participant_count.
    member_ids: list[uuid.UUID] = []
    participant_count: int = 0
    score: float | None = None


class ChallengeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    # Both null for an `open` challenge, which belongs to the platform, not a community.
    community_id: uuid.UUID | None
    # Names are included alongside the ids because the cross-community "my challenges" list
    # can't resolve them locally — a challenge there may belong to any community the caller
    # is in. Same rationale as MemeOut's community badge carrying a name.
    community_name: str | None = None
    opponent_community_id: uuid.UUID | None
    opponent_community_name: str | None = None
    # `open` challenges only: the reserved entry tag (normalized slug, no leading '#').
    hashtag: str | None = None
    creator: PublicUserOut
    # `duel` only: the challenged friend. Set from proposal time, before they've accepted
    # (and therefore before any `ChallengeParticipant` row for them exists).
    invitee_id: uuid.UUID | None = None
    invitee: PublicUserOut | None = None
    title: str
    challenge_type: ChallengeType
    status: ChallengeStatus
    start_time: datetime.datetime
    end_time: datetime.datetime
    winning_side_id: uuid.UUID | None
    sides: list[ChallengeSideOut]


class ChallengeSubmissionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    side_id: uuid.UUID
    submitter: PublicUserOut
    meme: MemeOut
    created_at: datetime.datetime


class ChallengeResultsOut(BaseModel):
    challenge: ChallengeOut
    submissions: list[ChallengeSubmissionOut]
